### ISSUE TITLE
Doesn't remove unnecessary parens around elements in type unions

### ISSUE THREAD

**Issue by mrapogee** [prettier/prettier#3939]
**Prettier 1.10.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEMCeAHOACeBnGbAXmwAooBXAWwCM4AnASmwB8yD6BLKAc1ewCGUCDAAWDRgG4QAGhAQMMTtDzJQA+vQgB3AAoaEqlAIBuETgBNZIAQWQAzAQBs8cOTXoCwAazgwAyhhe3DzIMPQUbiAWEGAOzq5y3K70MLqePFQC8S5RAFZ4AB4AQp4+fv4CVHAAMtxwOYnyFDAYLQBMjVFB9CnIqJhweGBcitYYXLAA6pZiyAAcAAxyExCuU54Y-RNDDCYNcvRwAI4UnEfpApnZSI65cq5UnGERUXghTnAAihQiDbcJKIwAQ0GYWOZIdpycICThOEIAYQgVCy-WEUAOIAorgAKiCjHdXABfIlAA)
```sh
--parser typescript
```

**Input:**
```tsx
type test = (number) | (string | another);
```

**Output:**
```tsx
type test = (number) | (string | another);

```

**Expected behavior:**
```tsx
type test = number | string | another
```

**Comment by j-f1** [comment]
With Flow, the parens around `(number)` are removed, but not the ones around `(string | another)`:

**Prettier 1.10.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEMCeAHOACeBnGbAXmwAooBXAWwCM4AnASmwB8yD6BLKAc1ewCGUCDAAWDRgG4QAGhAQMMTtDzJQA+vQgB3AAoaEqlAIBuETgBNZIAQWQAzAQBs8cOTXoCwAazgwAyhhe3DzIMPQUbiAWEGAOzq5y3K70MLqePFQC8S5RAFZ4AB4AQp4+fv4CVHAAMtxwOYnyFDAYLQBMjVFB9CnIIDQCNGhO0NYYXLAA6pZiyAAcAAxyExCuU54Y-RNwKSYNcvRwAI4UnEfpApnZSI65cq5UnGERUXghTnAAihQiDbcJKIwIYzCxzJDtOThAScJwhADCECoWX6wigBxAFFcABUhkY7q4AL6EoA)

**Input:**
```ts
type test = (number) | (string | another);
```

**Output:**
```ts
type test = number | (string | another);

```

**Comment by lydell** [comment]
I _think_ that the parens around `number` and the parens around `string | another` are two different cases.

The first one is parsed as a `TSParenthesizedType`. I don't know TypeScript enough to know what that is. Flow does not seem to have that concept.

The second case is interesting. It applies to Flow as well. The right-hand side parses as:

```json
{
  "type": "TSUnionType",
  "types": [
    {
      "type": "TSNumberKeyword",
    },
    {
      "type": "TSParenthesizedType",
      "typeAnnotation": {
        "type": "TSTypeAnnotation",
        "typeAnnotation": {
          "type": "TSUnionType",
          "types": [
            {
              "type": "TSStringKeyword",
            },
            {
              "type": "TSTypeReference",
              "typeName": {
                "type": "Identifier",
                "name": "another"
              }
            }
          ]
        }
      }
    }
  ]
}
```

If we were to remove those parens, we would change the AST, which Prettier avoids in general:

```json
{
  "type": "TSUnionType",
  "types": [
    {
      "type": "TSNumberKeyword",
    },
    {
      "type": "TSStringKeyword",
    },
    {
      "type": "TSTypeReference",
      "typeName": {
        "type": "Identifier",
        "name": "another"
      }
    }
  ]
}
```

In other words, it goes from being a union type with two elements, one of which is a nested union type, to a single flat union type with three elements.

I don't know those to are generally equivalent. Perhaps @vjeux knows (somebody who knows)?

**Comment by mrapogee** [comment]
I do think the cases are technically equivalent but I noticed:

```
test || (another || third)
```

becomes

```
test || (another || third);
```

This must be a recent change, because I remember there being cases when prettier would remove all unnecessary operator parens in math expressions, to the detriment of readability in some cases.

I'll adjust to just the first case `(number) | (string) | (another)`

**Comment by lydell** [comment]
@mrapogee Operators don't work the same. Remember, there are technically no sequences of operators, only left and right sides where each side in turn may be a nested operator expression.

**Prettier 1.10.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBDABAH0+gRlnMAbgB0oAKDbPASgPWLKp3P2rBqJABoQIAHGAEtoAZ2ShUAJykQA7gAVpCcSlQA3CEIAmPNKJjIAZqgA2ouL1xTUYANZwYAZX62hUAObIYUgK6WQbQgwYzMLXncLKRgFGw8AW1RQ8wCAK1EADwAhG3tHJ1R4uAAZdzhk8L5fGH5qgCYKgNcpKOQQXFRcAE9TaD1+KXcYAHUdGAALZAAOAAZeAYgLYZt+NoG4KPVy3ik4AEdfIV3Y1ASkpBMU3gt4oW8-ANF3D1M4AEVfCHhG3hhO0e0E2QdV+NiEpmeAGEIPFEm0oNBtiBfBYACqdVSXCwAX2xQA)

**Input:**
```jsx
a || b || c;
(a || b) || c;
a || (b || c);
```

**Output:**
```jsx
a || b || c;
a || b || c;
a || (b || c);

```

What happens above is that the first two cases parse to the same AST, so they are printed the same way. In the last case the parens affect the operator rules so it parses to a different tree. Logically, the mean the same thing, but I'm unsure if this could ever change the meaning of a program. _If_ Prettier were to remove the parens there, it would be in the same category as your type union case: The program would probably remain the same, but the AST would have changed.

**Comment by lydell** [comment]
I can't come up with a case that breaks the `||` or `&&` operators, but I _can_ come up with one for `+`.

One would think that `a + (b + c)` can be simplified to `a + b+ c`, but that is not safe since `(0.1 + 0.2) + 0.3 !== 0.1 + (0.2 + 0.3)` is a true statement in JavaScript.

So perhaps it's best not to muck with the AST in these cases.
